### PR TITLE
Add integration test for templates

### PR DESCRIPTION
**Pull request by drwpow**
## Changes

Partially addresses #365. Adds testing for the **local branch version** of every template, for both dev and build.

But it’s a partial fix as it’s also using the local copy of `create-astro` (i.e. monorepo symlinked). So it won’t catch the dependency misses that #365 describes for `npm init astro`. That will have to be handled separately in a followup PR.

It adds at least 2 minutes runtime (on a fast Macbook Pro) to the test suite because we are doing full npm installs and cold dev server startups. Feedback welcome there on how to improve it.

Screen Shot 2021-06-09 at 18 26 01

⚠️ **Note**: this PR isn’t set to run these template tests on every PR, mostly due to slowness. When/how should we run these?

## Testing

- [x] Tests are passing
- [x] Tests updated where necessary

## Docs

- [x] Docs / READMEs updated
- [x] Code comments added where helpful

### CODE REVIEW

File: packages/create-astro/src/index.ts [hunk 1]
```diff
@@ -52,12 +54,14 @@ export async function main() {
     },
   ]);
 
-  const emitter = degit(`snowpackjs/astro/examples/${options.template}`, {
+  const hash = args.commit ? `#${args.commit}` : '';
```
Comment (drwpow) [inline packages/create-astro/src/index.ts:57]: Adds a `--commit [hash]` flag to `create-astro`. This was necessary so we’re pulling PR commit hashes rather than `main`.

File: packages/create-astro/src/index.ts [hunk 2]
```diff
@@ -68,6 +72,22 @@ export async function main() {
...
 
+  // Post-process in parallel
+  await Promise.all(
+    POSTPROCESS_FILES.map(async (file) => {
+      const fileLoc = path.join(cwd, file);
+
+      switch (file) {
+        case 'package.json': {
+          const packageJSON = JSON.parse(await fs.promises.readFile(fileLoc, 'utf8'));
+          delete packageJSON.snowpack; // delete snowpack config only needed in monorepo (can mess up projects)
```
Comment (drwpow) [inline packages/create-astro/src/index.ts:83]: Unrelated fix: we have some artifacts in `package.json` from our monorepo since we decided to use the internal examples as public templates. This cleans up the `snowpack` setting which may break projects.

File: packages/create-astro/test/create-astro.test.js [hunk 3]
```diff
@@ -1,46 +1,120 @@
...
+      assert.not.ok(fs.existsSync(path.join(templateDir, file)), `failed to clean up ${file}`);
+    }
+  });
+
+  CreateAstro(`${template} (dev)`, async () => {
+    // start dev server
+    const port = 3000 + n; // start new port per test
+    const devServer = execa('yarn', ['start', '--port', port], { cwd: templateDir });
+    await new Promise((resolve) => {
+      setTimeout(() => resolve(), 15000);
```
Comment (drwpow) [inline packages/create-astro/test/create-astro.test.js:90]: Since we didn’t want to use a full browser test suite, we’re just giving the dev server 15 seconds to start up before running the test.

File: packages/create-astro/test/create-astro.test.js [hunk 4]
```diff
@@ -1,46 +1,120 @@
...
+    }); // give dev server flat 15s to set up
+    // TODO: try to ping dev server ASAP rather than waiting flat 15s
+
+    // ping dev server
+    const { statusCode, body } = await fetch(`http://localhost:${port}`);
+
+    // expect 200 to be returned with some response
+    assert.equal(statusCode, 200, 'didn’t respond with 200');
```
Comment (drwpow) [inline packages/create-astro/test/create-astro.test.js:98]: Dev server test 1: `200` response

File: packages/create-astro/test/create-astro.test.js [hunk 5]
```diff
@@ -1,46 +1,120 @@
...
+    assert.ok(body, 'returned empty response');
```
Comment (drwpow) [inline packages/create-astro/test/create-astro.test.js:99]: Dev server test 2: `body` returned something (i.e. not empty)

File: packages/create-astro/test/create-astro.test.js [hunk 6]
```diff
@@ -1,46 +1,120 @@
...
+
+    // clean up
+    devServer.kill();
+  });
+
+  CreateAstro(`${template} (build)`, async () => {
+    const MUST_HAVE_FILES = ['index.html', '_astro'];
```
Comment (drwpow) [inline packages/create-astro/test/create-astro.test.js:106]: Build test: we assert that a few files made it to `dist` (these will differ between templates, obviously, so just test a list of common paths for each)

File: tools/astro-vscode/package.json [hunk 7]
```diff
@@ -24,7 +24,7 @@
     "onLanguage:astro"
   ],
   "dependencies": {
-    "astro-languageserver": "file:../astro-languageserver"
```
Comment (drwpow) [inline tools/astro-vscode/package.json:27]: Fixes a `changeset` warning

Comment (drwpow) [inline tools/astro-vscode/package.json:27]: Fixes a changesets warning about this being a problem (if we need this I can revert it, but this probably shouldn’t get published, right?)

File: packages/create-astro/test/create-astro.test.js [hunk 8]
```diff
@@ -1,46 +1,120 @@
...
-for (const { value: template } of TEMPLATES) {
-  // TODO: Unskip once repo is made public. Because the repo is private, the templates can't yet be downloaded.
-  CreateAstro.skip(template, async () => {
-    const testDirectory = path.join(fixturesDir, template);
-    const { stdout } = await execa('../../create-astro.mjs', [testDirectory, '--template', template, '--force-overwrite'], { cwd: path.join(cwd, 'fixtures') });
+// enforce MAX_TEST_TIME
+CreateAstro.before.each(({ __test__ }) => {
+  if (TIMER[__test__]) throw new Error(`Test "${__test__}" already declared`);
+  TIMER[__test__] = setTimeout(() => {
+    throw new Error(`"${__test__}" did not finish within allowed time`);
```
Comment (drwpow) [inline packages/create-astro/test/create-astro.test.js:59]: Because `uvu` won’t ever finish if there are open threads, this test is super probable to run indefinitely. We had to add the same manual timeout handling here as we do in the astro tests.

**Review by matthewp — APPROVED** [review]
lgtm
